File: python/rainbow_test.py
from math import log2
import numpy as np
from sys import argv

from ver.effver import generate_random_linear_transformation, offVer
from ver.progver import progVer
from ver.ver import Ver
from ver.utils import fill_matrix_randomly

if len(argv) < 2 or argv[1] not in {'1', '2', '3'}:
    print('Defaulting to RainbowI.')

    from schemes.rainbow import RainbowI as Rainbow
    from schemes.gf.gf16 import gf16 as gf

    q = 16
elif argv[1] == '1':
    from schemes.rainbow import RainbowI as Rainbow
    from schemes.gf.gf16 import gf16 as gf

    q = 16
elif argv[1] == '2':
    from schemes.rainbow import RainbowIII as Rainbow
    from schemes.gf.gf256 import gf256 as gf

    q = 256
elif argv[1] == '3':
    from schemes.rainbow import RainbowV as Rainbow
    from schemes.gf.gf256 import gf256 as gf

    q = 256

pk_path = '/home/torres/Desktop/Thesis/verification_implementation/tmp/pk.txt'
signature_path = '/home/torres/Desktop/Thesis/verification_implementation/tmp/signature.txt'
message_path = '/home/torres/Desktop/Thesis/verification_implementation/tmp/debug.gdb'

PK = Rainbow.parse_public_key(pk_path)
s, salt = Rainbow.parse_signature(signature_path)
message_digest = Rainbow.message_digest(message_path, salt)
u = Rainbow.get_result_vector(message_digest)

print(f'VERIFICATION: {Ver(PK, s, u)}')

# for efficient verification
k = 1 #number of rows for the short verification key
C = generate_random_linear_transformation(k, PK.shape[0], 0, q-1, gf)
svk = C.dot(PK)
u_eff = C.dot(u)

assert(Ver(svk, s, u_eff))
print('Efficiently verified the Rainbow signature.')


# for progressive verification
t = 1

assert(progVer(PK, s, u, t))
print('Progressively verified the Rainbow signature.')


# for efficient and progressive verification
assert(progVer(svk, s, u_eff, t))
print('Efficiently and progressively verified the Rainbow signature.')


signature_guess = np.empty(PK.shape[1], dtype=object)

SAMPLE_SIZE = 100
eff_error_count = prog_error_count = effprog_error_count = 0

for _ in range(SAMPLE_SIZE):
    fill_matrix_randomly(signature_guess, min_value = 0, max_value = q-1, elem_type = gf)

    standard_ver_result = None

    if Ver(svk, signature_guess, u_eff):
        if standard_ver_result is None:
            standard_ver_result = Ver(PK, signature_guess, u)

        if not standard_ver_result:
            eff_error_count += 1
    
    if progVer(PK, signature_guess, u, t):
        if standard_ver_result is None:
            standard_ver_result = Ver(PK, signature_guess, u)

        if not standard_ver_result:
            prog_error_count += 1

    if progVer(svk, signature_guess, u_eff, t):
        if standard_ver_result is None:
            standard_ver_result = Ver(PK, signature_guess, u)

        if not standard_ver_result:
            effprog_error_count += 1

print('EFFICIENT VERIFICATION STATS')
print(f'Error percentage: {eff_error_count / SAMPLE_SIZE * 100}%')
print(f'Security bits: {0 if eff_error_count == 0 else log2(eff_error_count / SAMPLE_SIZE)}\n')

print('PROGRESSIVE VERIFICATION STATS')
print(f'Error percentage: {prog_error_count / SAMPLE_SIZE * 100}%')
print(f'Security bits: {0 if prog_error_count == 0 else log2(prog_error_count / SAMPLE_SIZE)}\n')

print('EFFICIENT+PROGRESSIVE VERIFICATION STATS')
print(f'Error percentage: {effprog_error_count / SAMPLE_SIZE * 100}%')
print(f'Security bits: {0 if effprog_error_count == 0 else log2(effprog_error_count / SAMPLE_SIZE)}')

File: c/test/rainbow_test.c
#include <assert.h>
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#include "../include/random_utils.h"
#include "../include/rainbow.h"
#include "../include/standard_verification.h"
#include "../include/efficient_verification.h"
#include "../include/progressive_verification.h"

#define STR_IMPL_(x) #x
#define STR(x) STR_IMPL_(x)

char const *pk_path = "/home/torres/Desktop/Thesis/verification_implementation/tmp/pk" STR(RAINBOW_VERSION) ".txt";
char const *signature_path = "/home/torres/Desktop/Thesis/verification_implementation/tmp/signature"  STR(RAINBOW_VERSION) ".txt";
char const *message_path = "/home/torres/Desktop/Thesis/verification_implementation/tmp/debug.gdb";

int main(int argc, char *argv[]) {

    if (argc != 3) {
        fprintf(stderr, "Wrong number of arguments.\n");
        exit(-1);
    }

    uint16_t k = strtoul(argv[1], NULL, 10);
    uint16_t t = strtoul(argv[2], NULL, 10);
    
    if (k > n_polynomials) {
        fprintf(stderr, "The number of rows of the SVK (%u) can't be bigger than the number of rows of the PK (%u)\n", k, n_polynomials);
        return -1;
    } else if (t > k) {
        fprintf(stderr, "The number of progressive steps (%u) can't be bigger than the number of rows of the SVK (%u)\n", t, k);
        return -1;
    }

    //srand((unsigned int) time(NULL)); maybe put it back if needed

    unsigned char salt[SALT_SIZE];

    MatrixDS* PK = parse_public_key(pk_path);
    MatrixDS* s = parse_signature(signature_path, salt);
    MatrixDS* u = get_result_vector(message_path, salt);
    
    assert(verify_signature_with_result_vector(PK, s, u));

    printf("Successfully verified the Rainbow signature.\n");
    
    // for efficient verification
    MatrixDS *C = CreateMatrix(k, nrows(PK), false);
    MatrixDS *SVK = offVer(PK, C, k);
    MatrixDS *u_eff = dot_product(C, u);

    assert(verify_signature_with_result_vector(SVK, s, u_eff));
    printf("Efficiently verified the Rainbow signature.\n");

    // for progressive verification
    assert(progVer_with_result_vector(PK, s, u, t, true));
    printf("Progressively verified the Rainbow signature.\n");

    assert(progVer_with_result_vector(SVK, s, u_eff, t, true));
    printf("Efficiently and progressively verified the Rainbow signature.\n");

    destroy_matrix(PK);
    destroy_matrix(s);
    destroy_matrix(u);
    
    destroy_matrix(C);
    destroy_matrix(SVK);
    destroy_matrix(u_eff);

    return 0;
}
